pennylane_lightning/src/gates/cpu_kernels/avx_common/ApplyIsingXX.hpp

```diff
@@ -0,0 +1,175 @@
+// Copyright 2022 Xanadu Quantum Technologies Inc.
+
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+
+//     http://www.apache.org/licenses/LICENSE-2.0
+
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+/**
+ * @file
+ * Defines IsingXX gate
+ */
+#pragma once
+#include "AVXUtil.hpp"
+#include "BitUtil.hpp"
+#include "Permutation.hpp"
+#include "Util.hpp"
+
+#include <immintrin.h>
+
+#include <complex>
+
+namespace Pennylane::Gates::AVX {
+
+template <typename PrecisionT, size_t packed_size> struct ApplyIsingXX {
+    using Precision = PrecisionT;
+    using PrecisionAVXConcept =
+        typename AVXConcept<PrecisionT, packed_size>::Type;
+
+    constexpr static size_t packed_size_ = packed_size;
+    constexpr static bool symmetric = true;
+
+    template <size_t rev_wire0, size_t rev_wire1>
+    static constexpr auto permutationInternalInternal() {
+        std::array<uint8_t, packed_size> perm = {
+            0,
+        };
+
+        size_t m = (1U << rev_wire0) | (1U << rev_wire1);
+        for (size_t k = 0; k < packed_size / 2; k++) {
+            perm[2 * k + 0] = 2 * (k ^ m) + 1;
+            perm[2 * k + 1] = 2 * (k ^ m) + 0;
+        }
+        return Permutation::compilePermutation<PrecisionT>(perm);
+    }
+
+    template <size_t rev_wire0, size_t rev_wire1, class ParamT>
+    static void applyInternalInternal(std::complex<PrecisionT> *arr,
+                                      size_t num_qubits, bool inverse,
+                                      ParamT angle) {
+        // This function is allowed for AVX512 and AVX2 with float
+
+        const double isin =
```

Comment:
As mentioned by @mlxd, could we use auto to account for both float & double parameters? Or we need to put it as float as mentioned in L56.

Reply:
:+1: